File: patch.go
package main

import (
	"sort"
	"strconv"
	"strings"
)

const (
	fileStatusAdded    = "added"
	fileStatusDeleted  = "deleted"
	fileStatusModified = "modified"
	fileStatusRenamed  = "renamed"
)

type PatchFile struct {
	Path     string `json:"path"`
	PrevPath string `json:"prevPath,omitempty"`
	Status   string `json:"status"`
}

func (file PatchFile) displayPath() string {
	return coalesce(file.Path, file.PrevPath)
}

func (file PatchFile) contextPaths() (string, string) {
	oldPath := coalesce(file.PrevPath, file.Path)
	newPath := coalesce(file.Path, oldPath)
	return oldPath, newPath
}

func coalesce(values ...string) string {
	for _, value := range values {
		if value != "" {
			return value
		}
	}
	return ""
}

func parsePatchFiles(patch string) []PatchFile {
	var files []PatchFile
	current := PatchFile{}
	hasCurrent := false

	flush := func() {
		if !hasCurrent {
			return
		}
		if current.Path == "" {
			current.Path = current.PrevPath
		}
		if current.Status == "" {
			current.Status = fileStatusModified
		}
		if current.Path != "" {
			files = append(files, current)
		}
		current = PatchFile{}
		hasCurrent = false
	}

	for _, line := range strings.Split(patch, "\n") {
		switch {
		case strings.HasPrefix(line, "diff --git "):
			flush()
			hasCurrent = true
			prev, next := parseDiffGitLine(line)
			current.PrevPath = prev
			current.Path = next
			current.Status = fileStatusModified
		case hasCurrent && strings.HasPrefix(line, "new file mode "):
			current.Status = fileStatusAdded
		case hasCurrent && strings.HasPrefix(line, "deleted file mode "):
			current.Status = fileStatusDeleted
		case hasCurrent && strings.HasPrefix(line, "rename from "):
			current.PrevPath = normalizePatchPath(strings.TrimPrefix(line, "rename from "))
			current.Status = fileStatusRenamed
		case hasCurrent && strings.HasPrefix(line, "rename to "):
			current.Path = normalizePatchPath(strings.TrimPrefix(line, "rename to "))
			current.Status = fileStatusRenamed
		case strings.HasPrefix(line, "--- "):
			if !hasCurrent {
				hasCurrent = true
			}
			prev := normalizePatchHeaderPath(strings.TrimPrefix(line, "--- "))
			if prev != "" {
				current.PrevPath = prev
			}
		case strings.HasPrefix(line, "+++ "):
			if !hasCurrent {
				hasCurrent = true
			}
			next := normalizePatchHeaderPath(strings.TrimPrefix(line, "+++ "))
			if next != "" {
				current.Path = next
			}
		}
	}
	flush()

	return dedupePatchFiles(files)
}

func parseDiffGitLine(line string) (string, string) {
	rest := strings.TrimPrefix(line, "diff --git ")
	prev, next, ok := splitDiffGitPaths(rest)
	if !ok {
		return "", ""
	}
	return normalizePrefixedPatchPath(prev), normalizePrefixedPatchPath(next)
}

func normalizePatchHeaderPath(path string) string {
	path = strings.TrimSpace(path)
	if idx := strings.IndexByte(path, '\t'); idx >= 0 {
		path = path[:idx]
	}
	if path == "/dev/null" {
		return ""
	}
	return normalizePrefixedPatchPath(path)
}

func normalizePatchPath(path string) string {
	path = strings.TrimSpace(path)
	if unquoted, err := strconv.Unquote(path); err == nil {
		path = unquoted
	} else {
		path = strings.Trim(path, "\"")
	}
	return path
}

func normalizePrefixedPatchPath(path string) string {
	path = normalizePatchPath(path)
	path = strings.TrimPrefix(path, "a/")
	path = strings.TrimPrefix(path, "b/")
	return path
}

func splitDiffGitPaths(rest string) (string, string, bool) {
	rest = strings.TrimSpace(rest)
	if rest == "" {
		return "", "", false
	}
	if strings.HasPrefix(rest, "\"") {
		prev, remaining, ok := splitQuotedPath(rest)
		if !ok {
			return "", "", false
		}
		next := strings.TrimSpace(remaining)
		if next == "" {
			return "", "", false
		}
		if strings.HasPrefix(next, "\"") {
			quotedNext, trailing, ok := splitQuotedPath(next)
			if ok && strings.TrimSpace(trailing) == "" {
				return prev, quotedNext, true
			}
		}
		return prev, next, true
	}
	return splitUnquotedDiffGitPaths(rest)
}

func splitQuotedPath(value string) (string, string, bool) {
	escaped := false
	for idx := 1; idx < len(value); idx++ {
		switch {
		case escaped:
			escaped = false
		case value[idx] == '\\':
			escaped = true
		case value[idx] == '"':
			return value[:idx+1], value[idx+1:], true
		}
	}
	return "", "", false
}

func splitUnquotedDiffGitPaths(rest string) (string, string, bool) {
	firstPrev := ""
	firstNext := ""
	for idx := strings.Index(rest, " b/"); idx >= 0; {
		prev := rest[:idx]
		next := strings.TrimSpace(rest[idx+1:])
		if strings.HasPrefix(prev, "a/") && strings.HasPrefix(next, "b/") {
			if firstPrev == "" {
				firstPrev = prev
				firstNext = next
			}
			if normalizePrefixedPatchPath(prev) == normalizePrefixedPatchPath(next) {
				return prev, next, true
			}
		}
		nextStart := idx + len(" b/")
		nextIdx := strings.Index(rest[nextStart:], " b/")
		if nextIdx < 0 {
			break
		}
		idx = nextStart + nextIdx
	}
	return firstPrev, firstNext, firstPrev != ""
}

func dedupePatchFiles(files []PatchFile) []PatchFile {
	seen := make(map[string]int, len(files))
	out := make([]PatchFile, 0, len(files))
	for _, file := range files {
		key := file.displayPath()
		if key == "" {
			continue
		}
		if existing, ok := seen[key]; ok {
			out[existing] = file
			continue
		}
		seen[key] = len(out)
		out = append(out, file)
	}
	return out
}

// parsePatch parses the git patch in a single pass, returning the deduped file
// list and a map from each path to the raw per-file patch chunk it belongs to.
func parsePatch(patch string) ([]PatchFile, map[string]string) {
	chunks := splitGitFilePatches(patch)
	if len(chunks) == 0 {
		// No "diff --git" headers (plain unified diff); the per-file chunk map is
		// meaningless, so fall back to a whole-patch parse for the file list.
		return parsePatchFiles(patch), nil
	}

	files := make([]PatchFile, 0, len(chunks))
	filePatches := make(map[string]string)
	for _, chunk := range chunks {
		for _, file := range parsePatchFiles(chunk) {
			files = append(files, file)
			if file.Path != "" {
				filePatches[file.Path] = chunk
			}
			if file.PrevPath != "" {
				filePatches[file.PrevPath] = chunk
			}
		}
	}
	return dedupePatchFiles(files), filePatches
}

func splitGitFilePatches(patch string) []string {
	var patches []string
	start := -1
	for lineStart := 0; lineStart < len(patch); {
		lineEnd := strings.IndexByte(patch[lineStart:], '\n')
		if lineEnd == -1 {
			lineEnd = len(patch)
		} else {
			lineEnd += lineStart + 1
		}

		if strings.HasPrefix(patch[lineStart:lineEnd], "diff --git ") {
			if start >= 0 {
				patches = append(patches, patch[start:lineStart])
			}
			start = lineStart
		}
		lineStart = lineEnd
	}
	if start >= 0 {
		patches = append(patches, patch[start:])
	}
	return patches
}

func sortedCommentPaths(comments []ReviewComment) []string {
	seen := make(map[string]struct{}, len(comments))
	paths := make([]string, 0, len(comments))
	for _, comment := range comments {
		if comment.Path == "" {
			continue
		}
		if _, ok := seen[comment.Path]; ok {
			continue
		}
		seen[comment.Path] = struct{}{}
		paths = append(paths, comment.Path)
	}
	sort.Strings(paths)
	return paths
}

File: src/patch.rs
use std::collections::HashMap;

use serde::Serialize;

#[derive(Clone, Debug, Serialize)]
pub struct PatchFile {
    pub path: String,
    pub prev_path: Option<String>,
    pub status: FileStatus,
}

#[derive(Clone, Copy, Debug, Serialize)]
pub enum FileStatus {
    Added,
    Deleted,
    Modified,
    Renamed,
}

#[derive(Default)]
struct PendingPatchFile {
    path: Option<String>,
    prev_path: Option<String>,
    status: Option<FileStatus>,
}

impl PatchFile {
    pub fn display_path(&self) -> &str {
        if self.path.is_empty() {
            self.prev_path.as_deref().unwrap_or_default()
        } else {
            &self.path
        }
    }

    pub fn context_paths(&self) -> (&str, &str) {
        let old_path = self.prev_path.as_deref().unwrap_or(self.path.as_str());
        let new_path = if self.path.is_empty() {
            old_path
        } else {
            self.path.as_str()
        };
        (old_path, new_path)
    }
}

pub fn parse_patch(patch: &str) -> (Vec<PatchFile>, HashMap<String, String>) {
    let chunks = split_git_file_patches(patch);
    if chunks.is_empty() {
        return (parse_patch_files(patch), HashMap::new());
    }

    let mut files = Vec::new();
    let mut file_patches = HashMap::new();
    for chunk in chunks {
        for file in parse_patch_files(&chunk) {
            if !file.path.is_empty() {
                file_patches.insert(file.path.clone(), chunk.clone());
            }
            if let Some(prev_path) = &file.prev_path {
                file_patches.insert(prev_path.clone(), chunk.clone());
            }
            files.push(file);
        }
    }
    (dedupe_patch_files(files), file_patches)
}

pub fn parse_patch_files(patch: &str) -> Vec<PatchFile> {
    let mut files = Vec::new();
    let mut current = PendingPatchFile::default();

    for line in patch.lines() {
        if let Some(rest) = line.strip_prefix("diff --git ") {
            current.flush(&mut files);
            if let Some((prev, next)) = parse_diff_git_line(rest) {
                current.prev_path = Some(prev);
                current.path = Some(next);
            }
            current.status = Some(FileStatus::Modified);
        } else if line.starts_with("new file mode ") {
            current.prev_path = None;
            current.status = Some(FileStatus::Added);
        } else if line.starts_with("deleted file mode ") {
            current.path = None;
            current.status = Some(FileStatus::Deleted);
        } else if line.starts_with("rename from ") {
            current.prev_path = normalize_patch_path(line.trim_start_matches("rename from "));
            current.status = Some(FileStatus::Renamed);
        } else if line.starts_with("rename to ") {
            current.path = normalize_patch_path(line.trim_start_matches("rename to "));
            current.status = Some(FileStatus::Renamed);
        } else if let Some(rest) = line.strip_prefix("--- ") {
            current.prev_path = normalize_patch_header_path(rest);
        } else if let Some(rest) = line.strip_prefix("+++ ") {
            current.path = normalize_patch_header_path(rest);
        }
    }
    current.flush(&mut files);

    dedupe_patch_files(files)
}

impl PendingPatchFile {
    fn flush(&mut self, files: &mut Vec<PatchFile>) {
        let prev_path = self.prev_path.take();
        let Some(path) = self.path.take().or_else(|| prev_path.clone()) else {
            *self = Self::default();
            return;
        };
        files.push(PatchFile {
            path,
            prev_path,
            status: self.status.take().unwrap_or(FileStatus::Modified),
        });
        *self = Self::default();
    }
}

fn parse_diff_git_line(rest: &str) -> Option<(String, String)> {
    let (prev, next) = split_diff_git_paths(rest)?;
    Some((
        normalize_prefixed_patch_path(&prev)?,
        normalize_prefixed_patch_path(&next)?,
    ))
}

fn normalize_patch_header_path(path: &str) -> Option<String> {
    let path = path.split('\t').next().unwrap_or_default().trim();
    if path == "/dev/null" {
        return None;
    }
    normalize_prefixed_patch_path(path)
}

fn normalize_patch_path(path: &str) -> Option<String> {
    let path = path.trim();
    let normalized = if path.len() >= 2 && path.starts_with('"') && path.ends_with('"') {
        unquote_patch_path(path)
    } else {
        path.trim_matches('"').to_string()
    };
    (!normalized.is_empty()).then_some(normalized)
}

fn normalize_prefixed_patch_path(path: &str) -> Option<String> {
    let path = normalize_patch_path(path)?;
    Some(
        path.strip_prefix("a/")
            .or_else(|| path.strip_prefix("b/"))
            .map(str::to_string)
            .unwrap_or(path),
    )
}

fn split_diff_git_paths(rest: &str) -> Option<(String, String)> {
    let rest = rest.trim();
    if rest.is_empty() {
        return None;
    }
    if rest.starts_with('"') {
        let (prev, remaining) = split_quoted_path(rest)?;
        let next = remaining.trim();
        if next.starts_with('"') {
            let (quoted_next, trailing) = split_quoted_path(next)?;
            if trailing.trim().is_empty() {
                return Some((prev, quoted_next));
            }
        }
        return Some((prev, next.to_string()));
    }
    split_unquoted_diff_git_paths(rest)
}

fn split_quoted_path(value: &str) -> Option<(String, String)> {
    let mut escaped = false;
    for (idx, ch) in value.char_indices().skip(1) {
        if escaped {
            escaped = false;
        } else if ch == '\\' {
            escaped = true;
        } else if ch == '"' {
            return Some((value[..=idx].to_string(), value[idx + 1..].to_string()));
        }
    }
    None
}

fn split_unquoted_diff_git_paths(rest: &str) -> Option<(String, String)> {
    let mut first = None;
    let mut search_start = 0;
    while let Some(relative_idx) = rest[search_start..].find(" b/") {
        let idx = search_start + relative_idx;
        let prev = rest[..idx].to_string();
        let next = rest[idx + 1..].trim().to_string();
        if prev.starts_with("a/") && next.starts_with("b/") {
            if first.is_none() {
                first = Some((prev.clone(), next.clone()));
            }
            if normalize_prefixed_patch_path(&prev) == normalize_prefixed_patch_path(&next) {
                return Some((prev, next));
            }
        }
        search_start = idx + " b/".len();
    }
    first
}

fn dedupe_patch_files(files: Vec<PatchFile>) -> Vec<PatchFile> {
    let mut seen = HashMap::new();
    let mut out = Vec::new();
    for file in files {
        let key = file.display_path().to_string();
        if key.is_empty() {
            continue;
        }
        if let Some(existing) = seen.get(&key).copied() {
            out[existing] = file;
        } else {
            seen.insert(key, out.len());
            out.push(file);
        }
    }
    out
}

fn split_git_file_patches(patch: &str) -> Vec<String> {
    let mut patches = Vec::new();
    let mut current = String::new();
    for line in patch.split_inclusive('\n') {
        if line.starts_with("diff --git ") && !current.is_empty() {
            patches.push(std::mem::take(&mut current));
        }
        current.push_str(line);
    }
    if !current.is_empty() && current.starts_with("diff --git ") {
        patches.push(current);
    }
    patches
}

fn unquote_patch_path(path: &str) -> String {
    let mut out = String::new();
    let mut chars = path[1..path.len() - 1].chars();
    while let Some(ch) = chars.next() {
        if ch != '\\' {
            out.push(ch);
            continue;
        }
        match chars.next() {
            Some('n') => out.push('\n'),
            Some('t') => out.push('\t'),
            Some('r') => out.push('\r'),
            Some('\\') => out.push('\\'),
            Some('"') => out.push('"'),
            Some(ch @ '0'..='7') => {
                let mut value = ch.to_digit(8).unwrap_or(0);
                for _ in 0..2 {
                    match chars.clone().next().and_then(|next| next.to_digit(8)) {
                        Some(digit) => {
                            chars.next();
                            value = value * 8 + digit;
                        }
                        None => break,
                    }
                }
                out.push(char::from_u32(value).unwrap_or('?'));
            }
            Some(other) => out.push(other),
            None => out.push('\\'),
        }
    }
    out
}
